File: src_qt/rules/ConditionPerceptionNode.cpp
#include <QtCore/QString>
#include <QtCore/QObject>
#include <QtCore/QListIterator>
#include <QDebug>

#include "ConditionPerceptionNode.h"
#include "../Brain.h"

namespace CnotiMind
{

	ConditionPerceptionNode::ConditionPerceptionNode( const QString& perception, const QString& value, ConditionOperator op, Brain* brain, QObject* parent):
		ConditionNode( op, brain, parent ),
		_perception( perception ),
		_value( value ),
		_valueNumeric( value.toInt( &_isValueNumeric ) )
	{

	}

	void ConditionPerceptionNode::exec()
	{
		if( isTrue() )
		{
			QHash<QString,QString> variables;

			// get the perception, being handled
			Perception p = _brain->_receivedPerceptions.head();

			variables.insert("[Perception.value]", p.value().toString() );
			execChildren( variables );
		}
	}

	void ConditionPerceptionNode::exec( QHash<QString,QString>& variables )
	{
		if( isTrue() )
		{
			// if the perception value is not in the variables list
			if( ! variables.contains( "[Perception.value]" ) )
			{
				// get the perception, being handled
				Perception p = _brain->_receivedPerceptions.head();

				// Insert the perception value in the variables
				variables.insert("[Perception.value]", p.value().toString() );
			}
			execChildren( variables );
		}
	}

	QString ConditionPerceptionNode::info( int depth ) const
	{
		QString info;

		info += space(depth) + "Condition type=Perception";

		return info + RuleNode::info(depth);
	}

	/*
		Test if there are perceptions, and test the first queued perception, if it is
		equal to this node perception.

		Optionally the value can be set, and the conditional operator.
	*/
	bool ConditionPerceptionNode::isTrue()
	{
		// Check if there are perceptions to be processed in the brain
		if( !_brain->_receivedPerceptions.isEmpty() )
		{
			// Just test the first perception
			Perception p = _brain->_receivedPerceptions.head();
			// Check if it is the percetion for this Node
			if( _perception.compare( p.name(), Qt::CaseInsensitive ) == 0 )
			{
				if( _value.isEmpty() )
				{
					return true;
				}
				else // test if it should be a perception with a specific value
				{
					bool ok, ok2;
					qreal perceptionValue = p.value().toDouble(&ok);
					qreal nodeValue = _value.toDouble(&ok2);

					if( ok && ok2 )
					{

						switch( _operator )
						{
							case ConditionOperatorBigger: return perceptionValue > nodeValue;
							case ConditionOperatorBiggerOrEqual: return perceptionValue >= nodeValue;
							case ConditionOperatorSmaller: return perceptionValue < nodeValue;
							case ConditionOperatorSmallerOrEqual: return perceptionValue <= nodeValue;
							case ConditionOperatorEqual: return perceptionValue == nodeValue;
							case ConditionOperatorDifferent: return perceptionValue != nodeValue;
						}
						return false;
					}
					else if( p.value().canConvert( QVariant::String ) )
					{
						// If it is a String value, it can only use equal or different operator.
						QString s = p.value().toString();
						switch( _operator )
						{
						case ConditionOperatorEqual:
							return QString::compare( _value, s, Qt::CaseInsensitive) == 0;
						case ConditionOperatorDifferent:
							return QString::compare( _value, s, Qt::CaseInsensitive) != 0;
						}
						return false;
					}
					else
					{
						// The value of the perceptions is not recognized,
						// so this node will return false
						return false;
					}

					// TODO, test the value using the condition Operator
				}
			}
		}

		return false;
	}

	ConditionPerceptionNode* ConditionPerceptionNode::fromXML(const QString &qName, const QXmlAttributes &atts, Brain *brain, QObject *parent)
	{
		if(qName.compare( "Condition", Qt::CaseInsensitive) == 0)
		{
			QString type = atts.value( "type" );
			if( type.compare("Perception", Qt::CaseInsensitive) == 0 )
			{

				QString perception = atts.value( "perception" );
				QString value = atts.value( "value" );
				ConditionOperator op = translateConditionOperator( atts.value( "operator" ) );


				return new ConditionPerceptionNode( perception, value, op, brain, parent );
			}
		}

		return NULL;
	}

}

File: src_ios/CnotiMind/ConditionPerceptionNode.m
//
//  ConditionEmotionNode.m
//  CnotiMind
//
//  Created by Gonçalo Rodrigues on 11/05/09.
//  Copyright 2011 Universidade de Coimbra. All rights reserved.
//

#import "ConditionPerceptionNode.h"

#import "Brain.h"
#import "Perception.h"

@implementation ConditionPerceptionNode


- (id) initWithPerceptionAndValueAndOperatorAndBrainAndParent: (NSString*)aKey value:(NSString*)aValue operator: (enum ConditionOperator)aOperator brain:(Brain*)aBrain parent:(id)aParent
{
    
    if (self == [super initWithKeyAndValueAndBrainAndParent: aKey value:aValue operator:aOperator brain:aBrain parent:aParent])
    {
    
    }
    return self;
}


- (void) exec
{
    
    if ([self isTrue]) {
        NSMutableDictionary* variables = [[NSMutableDictionary alloc] init];

        Perception* p = [[_brain receivedPerceptions] objectAtIndex:0];
        [variables setObject:[p value] forKey:@"[Perception.value]"];

        //  TODO
        [self execChildren:variables];
    }
}


- (void) exec:(NSMutableDictionary*)aVariables
{
    if ([self isTrue]) {
        // if the perception value is not in the variables list
        if ([aVariables objectForKey:@"[Perception.value]"]) {
            
            Perception* p = [[_brain receivedPerceptions] objectAtIndex:0];
            [aVariables setObject:[p value] forKey:@"[Perception.value]"];
        }
        
        [self execChildren: aVariables];
    }
}


- (NSString*) info:(int)aDepth
{
    NSString* info = [NSString stringWithFormat:@""];
    NSString* space = [self space:aDepth];
    
    info = [info stringByAppendingFormat:@"%@ Condition type=Perception perception=%@ value=%@", space, _key, _value];
    info = [info stringByAppendingFormat:@"%@", [super info:aDepth]];
    
    return info;
}


- (BOOL) isTrue
{
    // Check if there are perceptions to be processed in the brain
    
//    DLog(@"%d", [_brain.receivedPerceptions count]);
    
    if( [_brain.receivedPerceptions count]!=0 )
    {
        // Just test the first perception
        Perception* p = (Perception*)[_brain.receivedPerceptions objectAtIndex:0];
//        DLog(@"0 - %@ - %@",p.name, p.value);
//        DLog(@"0 - %@",_key);

        // Check if it is the percetion for this Node
        if( ![p.name caseInsensitiveCompare:_key] )
        {
//            DLog(@"1 - %@ - %@",p.name,p.value);
            
            if( [_value length]==0 )
            {
//                DLog(@"2 - %@ - %@",p.name,p.value);
                return true;
            }
            else // test if it should be a perception with a specific value
            {
                bool ok, ok2;
                
                ok = ([[NSScanner scannerWithString:[p value]] scanFloat:NULL]) ? TRUE : FALSE;                
                float perceptionValue = [[p value] floatValue];
                
                ok2 = ([[NSScanner scannerWithString:_value] scanFloat:NULL]) ? TRUE : FALSE;
                float nodeValue = [_value floatValue];
                
                if( ok && ok2 )
                {
                    switch( (int)_operator )
                    {
                        case ConditionOperatorBigger: 
                            return perceptionValue > nodeValue;
                        case ConditionOperatorBiggerOrEqual: 
                            return perceptionValue >= nodeValue;
                        case ConditionOperatorSmaller: 
                            return perceptionValue < nodeValue;
                        case ConditionOperatorSmallerOrEqual: 
                            return perceptionValue <= nodeValue;
                        case ConditionOperatorEqual: 
                            return perceptionValue == nodeValue;
                        case ConditionOperatorDifferent: 
                            return perceptionValue != nodeValue;
                    }
                    return false;
                }
                // TODO is String                
                else if([p value] != nil)
                {
                    // If it is a String value, it can only use equal or different operator.
                    NSString* s = [NSString stringWithFormat:@"%@",[p value]];
                    switch( (int)_operator )
                    {
						case ConditionOperatorEqual:
							return ![_value caseInsensitiveCompare:s];
						case ConditionOperatorDifferent:
							return [_value caseInsensitiveCompare:s];
                    }
                    return false;
                }
                else
                {
                    // The value of the perceptions is not recognized,
                    // so this node will return false
                    return false;
                }
                
                // TODO, test the value using the condition Operator
            }
        }
    }
    
    return false;
}


@end
